### ISSUE TITLE
Documentation: twisted.internet.reactor does not appear in the API docs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10753]
|<img alt="moof's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| moof reported|
|-|-|
|Trac ID|trac#1144|
|Type|defect|
|Created|2005-08-21 20:55:17Z|
|Branch|https://github.com/twisted/twisted/tree/twisted.internet.reactor-doc-1144|

Attachments:

* [reactor-doc-1144.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4b1/4b19229e380e9ca29673fdb967fd09cdcf4b5b44/b87ac6fae405d8e8ca6f56d1c3cc2d338fa4db7a.patch) (759 bytes) - added by collab on 2008-07-05 16:30:22Z - patch against r24237

<details><summary>Searchable metadata</summary>

```
trac-id__1144 1144
type__defect defect
reporter__moof moof
priority__highest highest
milestone__ 
branch__branches_twisted_internet_reactor_doc_1144 branches/twisted.internet.reactor-doc-1144
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1124657717000000 1124657717000000
changetime__1215877672000000 1215877672000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__moof cc__mwh cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="moof's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moof commented|
|-|-|

```
#!html
<pre>
Everywhere you look you see the line:

from twisted.internet import reactor

Yet there is no documentation for this module in the API, as it magically
removes itself.

Something you can click on that says:

* The default reactor is SelectReactor - with a link to it
* This alias works for any reactor that is installed, and is installed when the
reactor.install() method is called
* For more information see SelectReactor, the reactor basics page and the
various reactor interfaces (which should be listed in the help text)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Maybe there needs to be some mention of how the reactor works in the
introductory documentation.  Or maybe there is one already and it needs to be
highlighted.  Or maybe something else.

Note that twisted.internet.reactor is fully documented by the IReactor*
interfaces in twisted.internet.interfaces, it just may not be clear to people
new to Twisted.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="moof's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|moof commented|
|-|-|

```
#!html
<pre>
It's in the HOWTOs, in Choosing a Reactor, and in Reactor basics. But it's not
obvious, you have to do a bit of reading to realise this. 

It is also confusing. You're importing something the API docs say doesn't exist.
It should appear in the api docs, along with explanatory text.

See also issue 1145. This would be an ideal place to put that document, or refer
to it, at the very least.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Adding things to the API docs aren't as easy as you might think. Most of the
pre-canned Services for example don't have proper API docs, because they aren't
defined using standard Python class and def declarations.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="edsuom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|edsuom set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

maybe we can do something with TwistedModel or whatever in pydoctor, to put something at `twisted.internet.reactor`?

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_9"></a>|@mwhudson commented|
|-|-|

Well yes I guess so.  What would you like to see there?

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|collab commented|
|-|-|

Replying to [mwh](#note_9):
> Well yes I guess so. What would you like to see there?

I think moof made a good suggestion, the text is already [there](http://twistedmatrix.com/projects/core/documentation/howto/reactor-basics.html):

Replying to [moof](#note_3):
> It's in the HOWTOs, in Choosing a Reactor, and in Reactor basics. But it's not
> obvious, you have to do a bit of reading to realise this.

So something like:

```
The reactor is the core of the event loop within Twisted -- the loop which drives applications using Twisted.
The reactor provides basic interfaces to a number of services, including network communications, threading, and event dispatching.
```

**Comment by twisted-trac** [comment]
|<img alt="moof's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|moof commented|
|-|-|

At the very least, either a straight copy of IReactor, or something pointing to it, and a short phrase saying that the default is SelectReactor. If you're feeling generous, a link the the different possible reactors, and maybe the docs telling you how to choose one.

The phrase suggested by collab is not a bad intro, but not enough.

The big problem is that people use twisted.internet.reactor all over the place. New people to twisted, reading twisted code for the first time, or even the docs, will naturally want to look it up. And they'll find nothing. And this is confusing. A nice link to the Right Place will help wonders.

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|collab set owner to collab<br>collab set status to `assigned`|
|-|-|

Replying to [moof](#note_11):
> The big problem is that people use twisted.internet.reactor all over the place. New people to twisted, reading twisted code for the first time, or even the docs, will naturally want to look it up. And they'll find nothing. And this is confusing. A nice link to the Right Place will help wonders.

This is a very good point, enough to raise the priority of this docs ticket to the max imo. I'll work on a patch based on moof's suggestions, if someone has a better idea, let's hear it. :)

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|collab commented|
|-|-|

Replying to [collab](#note_12):
> ... enough to raise the priority of this docs ticket to the max imo.

.. especially when this ticket is > 3 years old! :)

**Comment by twisted-trac** [comment]
|<img alt="collab's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|collab removed owner<br>collab set status to `new`|
|-|-|

Attached a [[reactor-doc-1144.patch](None/reactor-doc-1144.patch) patch] that covers most of moof's points, up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Good start, and good idea - we don't really need pydoctor development to happen in order to resolve this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [24250]) Branching to 'twisted.internet.reactor-doc-1144'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [24251]) apply somewhat modified reactor-doc-1144.patch

refs [#10753](https://github.com/twisted/twisted/issues/10753)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

Updating for my nickname

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

Looks good to me, maybe it would be useful to use:

```
@see: L{IReactorUDP<twisted.internet.interfaces.IReactorUDP>} 
```

instead of

```
@see: L{twisted.internet.interfaces.IReactorUDP} 
```

for better readability in the generated docs?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [24261]) provider shorter names

refs [#10753](https://github.com/twisted/twisted/issues/10753)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set status to `closed`|
|-|-|

(In [24262]) Merge twisted.internet.reactor-doc-1144

Author: exarkun
Reviewer: thijs
Fixes: [#10753](https://github.com/twisted/twisted/issues/10753)

Add a module docstring to twisted.internet.reactor so that pydoctor has something
to pick up and include on the page for that module.  The docstring briefly explains
the reactor and points to the documentation for a number reactor interfaces.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
